src/game/stage/spot_light.c

```diff
@@ -0,0 +1,174 @@
+#include "global.h"
+#include "core.h"
+#include "task.h"
+#include "trig.h"
+
+#include "game/game.h"
+#include "game/stage/spot_light.h"
+#include "game/stage/unknown_task.h"
+
+/* Spot Lights that appear in the background of Ice Paradise
+   (seem to be the same as in Sonic Advance's Secret Base Zone) */
+
+typedef struct {
+    struct Task *t0;
+    struct Task *t1;
+    s16 unk8;
+    s32 unkC;
+} Struct_StageSpotLight;
+
+void Task_SpotLightMain(void);
+void Task_800A8E0(void);
+void TaskDestructor_SpotLightMain(struct Task *);
+
+bool32 CreateSpotLightBeams(void)
+{
+    struct Task *t = TaskCreate(Task_SpotLightMain, sizeof(Struct_StageSpotLight),
+                                0x2000, 0, TaskDestructor_SpotLightMain);
+    Struct_StageSpotLight *spotLight = TaskGetStructPtr(t);
+    Struct_StageUnkTask *ut;
+
+    spotLight->unk8 = 0x600;
+    spotLight->unkC = 0;
+
+    spotLight->t0 = CreateStageUnknownTask();
+
+    ut = TaskGetStructPtr(spotLight->t0);
+    ut->unk6 = 60;
+    ut->unk8 = 200;
+    ut->unk0 = 0;
+    ut->unk2 = 32;
+    ut->unk4 = 0x400;
+    ut->unkA = 2;
+    ut->unkB = 0;
+
+    spotLight->t1 = CreateStageUnknownTask();
+
+    ut = TaskGetStructPtr(spotLight->t1);
+    ut->unk6 = 200;
+    ut->unk8 = 240;
+    ut->unk0 = 0;
+    ut->unk2 = 16;
+    ut->unk4 = 0x300;
+    ut->unkA = 3;
+    ut->unkB = 0;
+
+    return TRUE;
+}
+
+void Task_SpotLightMain(void)
+{
+    Struct_StageSpotLight *spotLight = TaskGetStructPtr(gCurTask);
+    Struct_StageUnkTask *ut;
+    s32 unkC;
+
+    if (!(gUnknown_03005424 & EXTRA_STATE__100)) {
+        ut = TaskGetStructPtr(spotLight->t0);
+
+        if (((gUnknown_03005590 & 0x7) == 0) && (ut->unkB != 0)) {
+            if (gDispCnt & DISPCNT_BG0_ON) {
+                if (ut->unkB < 32)
+                    ut->unkB++;
+
+                ut = TaskGetStructPtr(spotLight->t1);
+
+                if (ut->unkB < 16)
+                    ut->unkB++;
+            } else {
+                goto goto_labl;
+            }
+        } else if (!(gDispCnt & DISPCNT_BG0_ON)) {
```

Comment:
If you add `&& (gDispCnt & DISPCNT_BG0_ON)` to the top level if, you may find that the match for the else go to is solved

Reply:
As the compiler will skip the `!(gDispCnt & DISPCNT_BG0_ON)` condition that it's already checked

Reply:
I might be doing something wrong, but that doesn't work for me.

Reply:
I will have a go

Reply:
Fixed